### PR TITLE
[LP#2067404] Check and log all events on failed resources

### PR DESCRIPTION
**Pull request by addyess**
[LP#2067404](https://bugs.launchpad.net/bugs/2067404)

## Overview

It's been reported the charm is stuck in a Waiting state that doesn't resolve due to conditions on the `csi-cinder-controllerplugin` deployment resource not going active.  These changes will log to the charm debug log in the event of such circumstances. 

## Details

Gather the deployment and pod events associated with the deployment (or daemonsets) deployed by this charm to determine if there is anything the juju administrator can do to improved the deployment state. 

Events will be logged in the charm's debug log in timestamp order, least recent to most recent.

### CODE REVIEW

File: src/storage_manifests.py [hunk 1]
```diff
@@ -145,6 +155,51 @@ def evaluate(self) -> Optional[str]:
                 return f"Storage manifests waiting for definition of {prop}"
         return None
 
+    def is_ready(self, obj: HashableResource, cond: AnyCondition) -> Optional[bool]:
```
Comment (addyess) [inline src/storage_manifests.py:163]: overrides the build in `Manifests:is_ready` method to add extra debugging

File: src/storage_manifests.py [hunk 2]
```diff
@@ -145,6 +155,51 @@ def evaluate(self) -> Optional[str]:
...
+        """Determine if the resource is ready."""
+        if not super().is_ready(obj, cond):
+            object_events = collect_events(self.client, obj.resource)
+            if obj.kind in ["Deployment", "DaemonSet"]:
+                involved_pods = self.client.list(
+                    Pod, namespace=obj.namespace, labels={"app": obj.name}
+                )
+                object_events += [
+                    event for pod in involved_pods for event in collect_events(self.client, pod)
+                ]
```
Comment (addyess) [inline src/storage_manifests.py]: find any pods which may be involved with this deployment or daemonset and add their events as well

File: src/storage_manifests.py [hunk 3]
```diff
@@ -145,6 +155,51 @@ def evaluate(self) -> Optional[str]:
...
+            for event in sorted(object_events, key=by_localtime):
+                log.info(
```
Comment (addyess) [inline src/storage_manifests.py]: i think info level is appropriate here

File: src/storage_manifests.py [hunk 4]
```diff
@@ -145,6 +155,51 @@ def evaluate(self) -> Optional[str]:
...
+                    "Event %s/%s %s msg=%s",
+                    event.involvedObject.kind,
+                    event.involvedObject.name,
+                    event.lastTimestamp
+                    and event.lastTimestamp.astimezone()
+                    or "Date not recorded",
```
Comment (addyess) [inline src/storage_manifests.py]: localize the timestamp -- but sometimes events don't have an associated timestamp so we'll just say "Date not recorded" on those

File: src/storage_manifests.py [hunk 5]
```diff
@@ -145,6 +155,51 @@ def evaluate(self) -> Optional[str]:
...
+    return list(
+        client.list(
+            Event,
+            namespace=resource.metadata.namespace,
+            fields={
+                "involvedObject.kind": kind,
+                "involvedObject.name": resource.metadata.name,
+            },
+        )
+    )
```
Comment (eaudetcobello) [inline src/storage_manifests.py:211]: Is this a list of lists?

Comment (addyess) [reply src/storage_manifests.py:211]: [lightkube `client.list` ](https://lightkube.readthedocs.io/en/stable/client/) returns an iterator.  So this is actually building an ACTUAL python list of events from the returned python iterator. 

iterators are something you can call `next(it)` on until it yields a `StopIteration` exception. 

so by running `list(it)` you get an in-memory list version of that iterated item.  similar to `list(range(10))` providing you `[0,1,2,3,4,5,6,7,8,9]`.  the `range(...)` method yields an iterator

**Comment by addyess** [comment]
@HomayoonAlimohammadi how does it look.  Are we happy?

**Comment by HomayoonAlimohammadi** [comment]
@addyess Yea I was waiting for the CI to pass. Let's merge this!

**Pull request merged**